### PR TITLE
OXXO APM Payment requests/responses

### PR DESCRIPTION
**Pull request by kevin-reece-cko**
- includes synchronisation of minor benefitpay changes

Preview:
- https://api-reference.checkout.com/preview/task/oxxo/

Note: 
- further useful fields will be added with https://checkout.atlassian.net/browse/DA-88

### CODE REVIEW

File: spec/components/schemas/Payments/ResponseSources/PaymentResponseOXXOSource.yaml [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+###
+type: object
+description: 'OXXO Source'
+allOf:
+  -
+    $ref: '#/components/schemas/PaymentResponseSource'
+  -
+    type: object
+    properties:
+      status_message:
```
Comment (mikhail-kuzmiankou-cko) [inline spec/components/schemas/Payments/ResponseSources/PaymentResponseOXXOSource.yaml]: Note, this will be changed to failure_reason and failure_code soonish

File: spec/components/schemas/Payments/ResponseSources/PaymentResponseOXXOSource.yaml [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+        enum:
+          - rejected_by_bank
+          - insufficient_amount
+          - score_validation
+          - max_attempts_reached
+          - duplicated_payment
+          - rejected_due_to_high_risk
+          - invalid_transaction
+          - amount_exceeded
+          - error_in_acquirer
```
Comment (mikhail-kuzmiankou-cko) [inline spec/components/schemas/Payments/ResponseSources/PaymentResponseOXXOSource.yaml]: List of the errors will be changed as well

File: spec/components/schemas/Payments/RequestSources/PaymentRequestOXXOSource.yaml [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    type: object
+    required:
+      - country
+      - payer
+    properties:
+      integration_type:
+        enum:
+          - redirect
+        type: string
+        description: 'The type of the integration.  Currently only `redirect`, for which a redirect url is returned,  which the merchant redirects the customer to, to continue payment'
```
Comment (mikhail-kuzmiankou-cko) [inline spec/components/schemas/Payments/RequestSources/PaymentRequestOXXOSource.yaml]: > `integration.  Currently`

Extra space character is inserted when there is a trailing space at the end of the line in the original spec file. We should not insert a space at line ends.

Comment (kevin-reece-cko) [reply spec/components/schemas/Payments/RequestSources/PaymentRequestOXXOSource.yaml]: It's actually partly intentional, but not working... Basically there is supposed to be a new line there and `redoc` (the api reference ui framework) is very picky for newlines, and after many attempts I've found that there needs to be `   \n` (three spaces!), to render a new line. So in the dlocal specs that means:
```
              description: >-
                The type of the integration.   

                Currently only `redirect`, for which a redirect url is
```

Comment (kevin-reece-cko) [reply spec/components/schemas/Payments/RequestSources/PaymentRequestOXXOSource.yaml]: I think it's worth adding the extra whitespace for the redoc newline, as redoc is our most public facing usage of the specs.

Sound good?

File: spec/components/schemas/Payments/ResponseSources/PaymentResponseOXXOSource.yaml (hunk 1 above)
Comment (kevin-reece-cko) [reply spec/components/schemas/Payments/ResponseSources/PaymentResponseOXXOSource.yaml]: updated now

File: spec/components/schemas/Payments/ResponseSources/PaymentResponseOXXOSource.yaml (hunk 2 above)
Comment (kevin-reece-cko) [reply spec/components/schemas/Payments/ResponseSources/PaymentResponseOXXOSource.yaml]: updated now

File: spec/components/schemas/Payments/RequestSources/PaymentRequestOXXOSource.yaml (hunk 3 above)
Comment (mikhail-kuzmiankou-cko) [reply spec/components/schemas/Payments/RequestSources/PaymentRequestOXXOSource.yaml]: If this is the only option, then it sounds good 🙂 YAML rules about indentation and line breaks are quite complex: https://stackoverflow.com/a/21699210. I wonder if we can find a mode that will be more intuitive.

File: spec/components/schemas/Payments/RequestSources/PaymentRequestOXXOSource.yaml [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        enum:
+          - redirect
+        type: string
+        description: "The type of the integration.   \nCurrently only `redirect`, for which a redirect url is returned,  which the merchant redirects the customer to, to continue payment"
+      country:
+        maxLength: 2
+        enum:
+          - MX
+        type: string
+        description: 'User''s country code. ISO 3166-1 alpha-2 code.'
```
Comment (mikhail-kuzmiankou-cko) [inline spec/components/schemas/Payments/RequestSources/PaymentRequestOXXOSource.yaml]: Should we rename 'user' to 'payer'?

File: spec/components/schemas/Payments/ResponseSources/PaymentResponseOXXOSource.yaml [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+description: 'OXXO Source'
+allOf:
+  -
+    $ref: '#/components/schemas/PaymentResponseSource'
+  -
+    type: object
+    properties:
+      failure_reason:
+        type: string
+        description: 'Error details'
```
Comment (mikhail-kuzmiankou-cko) [inline spec/components/schemas/Payments/ResponseSources/PaymentResponseOXXOSource.yaml]: Maybe rephrase to something like "Reason of payment failure"

File: spec/components/schemas/Payments/ResponseSources/PaymentResponseOXXOSource.yaml [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+          - third_party_access_denied
+          - third_party_invalid_request
+          - third_party_invalid_parameter
+          - third_party_invalid_transaction_status
+          - third_party_rejected
+          - third_party_amount_exceeded
+          - third_party_timeout
+          - third_party_error
+        type: string
+        description: 'Error code'
```
Comment (mikhail-kuzmiankou-cko) [inline spec/components/schemas/Payments/ResponseSources/PaymentResponseOXXOSource.yaml]: > Payment failure code

File: spec/components/schemas/Payments/RequestSources/PaymentRequestOXXOSource.yaml (hunk 3 above)
Comment (kevin-reece-cko) [reply spec/components/schemas/Payments/RequestSources/PaymentRequestOXXOSource.yaml]: Yeah, I often find myself at that stack overflow answer 😂 

The extra complexity comes from our mapping to json on the apm api, and then back to yaml for the api reference. Json specs don't seem to support the full complexity of yaml spec whitespace rules, so it gets compressed to a single line string in the json form, and then doesn't expand back to multi-line in yaml. 

So as Json doesn't support multi-line, we have to map to json as `(space)(space)(space)\n` or some other `YAML.stringify` compatible space/newline combo. Or expose yaml specs from the apms.

I will have another look at other possibilities for the yaml -> json and json -> yaml mappings

Comment (mikhail-kuzmiankou-cko) [reply spec/components/schemas/Payments/RequestSources/PaymentRequestOXXOSource.yaml]: Oh yeah, not trivial indeed 😱 Anyway, three spaces is fine 🙂

File: spec/components/schemas/Payments/RequestSources/PaymentRequestOXXOSource.yaml (hunk 4 above)
Comment (kevin-reece-cko) [reply spec/components/schemas/Payments/RequestSources/PaymentRequestOXXOSource.yaml]: probably, for consistency

**Comment by mikhail-kuzmiankou-cko** [comment]
@kevin-reece-cko Please port back the changes to checkout-ap-dlocal repo.

**Comment by mikhail-kuzmiankou-cko** [comment]
Nice, I see it is done in this PR: https://github.com/CKOTech/checkout-ap-dlocal/pull/27/files
